rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+
+*   keep the number of scheduled but not executed functions small. This may be difficult to achieve since there is not an easy way for the client to know whether a function is executed or not. The only way is to synchronize the executor. Therefore, as a workaround we will have to periodically synchronize the executor to make sure functions are actually executed, before the client schedules more functions. In the long run, we should get acknowledgement from runtime about how many functions have been executed.
+*   eagerly fetch the outputs of remote functions once the outputs are passed to `strategy.local_result`. In this way, we can know the status of function execution earlier.
+*   recommend users schedule only small functions. Large functions are more expensive to retry.
+
+
+### Evaluation
+
+Historically, `tf.estimator.Estimator` uses a dedicated evaluator that periodically loads from a checkpoint, and performs evaluation with evaluation data. However `tf.keras` typically evaluates in an alternating manner after every epoch of training, and this is also the case with `tf.keras` + `MultiWorkerMirroredStrategy`. 
+
+With `ParameterServerStrategyV2`, we will start with a dedicated** evaluator that runs alongside the training cluster**, **aka “sidecar evaluation”**; in this scheme, training client is required to generate checkpoints periodically, and the evaluator reads the latest checkpoint as it becomes available. The evaluation is asynchronous to the training progress. With this we provide the functionality Estimator has been able to with Keras API, which is important to attract updates from Estimator users to TF 2.0.
+
+With our recommendation, users should create a separate evaluation client that runs the same python binary as the training client. This python binary will contain the if-else clause as it bifurcates into two paths:
+
+
+```python
+if cluster_resolver.task_type == "chief":
+  run_training_loop()
+elif cluster_resolver.task_type == "evaluator":
+  run_evaluation_loop()
+```
+
+
+Evaluation code:
+
+
+```python
+def run_evaluation_loop(...):
+  """Run the example custom evaluation loop."""
+  
+  eval_dataset, model, eval_accuracy = ...
+  checkpoint = tf.train.Checkpoint(model=model)
+
+  @tf.function
+  def eval_fn(eval_dataset):
+    for _ in range(eval_steps):
+      # evaluation pass
+    return eval_accuracy.result()
+
+  while True:
+    latest_checkpoint = get_new_checkpoint()
+    checkpoint.restore(latest_checkpoint)
+    eval_result = eval_fn(iterator)  # Users can print, early stop, mark ckpt..
+```
+
+
+In the evaluation client, the user loads the checkpoints that were periodically saved into the model (by the training client), does evaluation over a full pass of eval dataset, and does whatever they want to do with eval results. Examples include exporting them to files which can be read by the training client for actions (such as reducing learning rate, early stopping, etc.)
+
```

Comment:
Can we see the e2e code example with training_loop and the fork as well?

Reply:
Yes - we have some code sample and will update in the doc.